[core] Add `token-auth.md` documentation page (#58829)
ray-project/ray: doc/source/ray-security/token-auth.md

The following example shows what happens on the first run:

```bash
$ export RAY_AUTH_MODE=token
$ python -c "import ray;ray.init()"
```

On the first run, this command (or any other script that initializes Ray) will log a line similar to the following:

```bash
Generated new authentication token and saved to /Users/<username>/.ray/auth_token
```

### Local development with ray start

When you use `ray start --head` to start a local cluster after setting `RAY_AUTH_MODE=token` as an environment variable, you need to generate a token first:

- If no token exists, `ray start` shows an error message with instructions.
- Run `ray get-auth-token --generate` to generate a new token at the path `~/.ray/auth_token`.
- Once generated, Ray uses the token every time you run `ray start`.

The following example demonstrates this flow:

```bash
# Set the environment variable.
$ export RAY_AUTH_MODE=token

# First attempt - an error is raised if no token exists.
$ ray start --head
...
RuntimeError: Token authentication is enabled but no authentication token was found. Please provide an authentication token using one of these methods:
  1. Set the RAY_AUTH_TOKEN environment variable

>   2. Set the RAY_AUTH_TOKEN_PATH environment variable (pointing to a token file)

Reviewer: I realise these are in the (correct) priority order, but do we want to suggest token path before token? It's probably better not to encourage people to litter the token in their environment

Author: agreed. this one requires code change -- added to it "post release todo list"

```bash
scp ~/.ray/auth_token user@node1:~/.ray/auth_token;
scp ~/.ray/auth_token user@node2:~/.ray/auth_token;
```

#### Step 3: Start Ray with token authentication

You must set the environment variable `RAY_AUTH_MODE=token` on each node before running `ray start`, as in the following example:

```bash
ssh user@node1 "RAY_AUTH_MODE=token ray start --head";
ssh user@node2 "RAY_AUTH_MODE=token ray start --address=node1:6379";
```

## Troubleshooting token authentication issues

You might encounter the following problems with token authentication.

### Token authentication isn't enabled

Make sure you've set the `RAY_AUTH_MODE=token` environment variable in the environment where you're launching Ray *and* in any shell where you are using a client to connect to Ray.

### Authentication token not found

If running locally, run `ray get-auth-token --generate` to create a token on your local machine.
If running a remote cluster, make sure you've followed instructions to copy your token into the cluster.

### Invalid authentication token

Any client that tries to interact with a Ray cluster must have the same token as the Ray cluster.

> If the token on your local machine doesn't match the token in a Ray cluster, you can use the `RAY_AUTH_TOKEN` environment variable to configure a token for interacting with that cluster. You must work with the creator of the cluster to get the token.

Reviewer: Should we steer people toward RAY_AUTH_TOKEN_PATH here as well?

Author: done, and added recommendation

> (token-auth)=

Author: nope, this defines the tag so the page can be cross linked without hardcoding its filepath

# Ray token authentication

Enable token authentication in Ray to secure cluster access and prevent unauthorized use. This guide explains how authentication works and how to set it up for different deployment scenarios.

:::{note}
Token authentication is available in Ray 2.52.0 or later.
:::

## How Ray token authentication works

To enable token authentication, set the environment variable `RAY_AUTH_MODE=token` before starting your Ray cluster. When you start a Ray cluster with authentication enabled, all external Ray APIs and internal communications are authenticated using the token as a shared secret.

> The process for generating and configuring authentication tokens differs depending on how you launch your Ray cluster. When you start a local instance of Ray using `ray.init()` with token authentication enabled, a token will automatically be generated and used.

Reviewer 2: ```suggestion
The process for generating and configuring authentication tokens differs depending on how you launch your Ray cluster. When you start a local instance of Ray using `ray.init()` with token authentication enabled, Ray auto automatically generates and uses a token.
```

- **TLS termination**: Deploy a TLS proxy in front of your Ray cluster.
- **VPN/Overlay networks**: Use network-level encryption for all traffic into and within the cluster.

:::{warning}
Don't expose Ray clusters directly to the internet without encryption. Tokens alone don't protect against network eavesdropping.
:::

Tokens have the following properties:

- Ray stores tokens by default in plaintext at `~/.ray/auth_token`.
- Use file permissions to keep token files secure, especially in shared environments.
- Don't commit tokens to version control.
- Tokens don't expire. Your local token remains valid unless you delete and regenerate the token. Ray clusters use the same token for the lifetime of the cluster.

## Configure token authentication for local development

To enable authentication on your local machine for development, set the `RAY_AUTH_MODE=token` environment variable in your shell or IDE. You can persist this configuration in your `.bashrc` file or similar.

### Local development with ray.init()

When you run a script containing `ray.init()` after setting `RAY_AUTH_MODE=token` as an environment variable, Ray handles authentication automatically:

> - If a token doesn't already exist at `~/.ray/auth_token`, Ray generates a token and saves it to the file. A log message displays to confirm token creation.

Reviewer 2: Just confirming that `address=auto` is the default?

        resources: {}
        node_config:
            InstanceType: m5.large

# Mount a locally generated token file to all nodes in the Ray cluster.
file_mounts: {
    "/home/ubuntu/.ray/auth_token": "~/.ray/auth_token",
}

# Set the RAY_AUTH_MODE environment variable for all shell sessions on the cluster.
initialization_commands:
  - echo "export RAY_AUTH_MODE=token" >> ~/.bashrc
```

#### Step 3: Launch the Ray cluster

Run the following command to launch a Ray cluster using your cluster configuration YAML:

```bash
ray up cluster.yaml
```

#### Step 4: Configure the Ray dashboard and port forwarding

Connecting to the Ray dashboard configures secure SSH port forwarding between your local machine and the Ray cluster. Complete this step even if you don't plan to use the dashboard for monitoring.

Run the following command to set up port forwarding for the Ray dashboard port (`8265` by default):

```bash
ray dashboard cluster.yaml
```

Upon opening the dashboard, a prompt displays requesting your authentication token. To display the token in plaintext, you can run the following on your local machine:

```bash
export RAY_AUTH_MODE=token
ray get-auth-token
```

> Paste the token in the prompt and click **Submit**. The token gets stored as a cookie for a maximum of 30 days. When you open the dashboard for a new cluster that uses a different token, you will be prompted to re-enter the token.

Reviewer 2: ```suggestion
Paste the token in the prompt and click **Submit**. The token gets stored as a cookie for a maximum of 30 days. When you open the dashboard for a cluster that uses a different token, a prompt appears to enter the token for that cluster.
```
